syft/pkg/license.go

```diff
@@ -0,0 +1,157 @@
+	"github.com/anchore/syft/syft/source"
+)
+
+var _ sort.Interface = (*Licenses)(nil)
+
+// License represents an SPDX Expression or license value extracted from a packages metadata
+// We want to ignore URL and Location since we will always merge these fields
+// a License is a unique combination of value, expression and type
+// its sources are always considered merge and additions to the evidence
+// of where it was found and how it was sourced.
+// This is different from how we treat a package since we consider package paths to distinguish
+// if packages should be kept separate, this is different for licenses since we're only looking for evidence
+// of where a license was declared/concluded for a package
+type License struct {
+	Value          string             `json:"value"`
+	SPDXExpression string             `json:"spdxExpression"`
+	Type           license.Type       `json:"type"`
+	URL            internal.StringSet `hash:"ignore"`
+	Location       source.LocationSet `hash:"ignore"`
+}
+
+type Licenses []License
+
+func (l Licenses) Len() int {
+	return len(l)
+}
+
+func (l Licenses) Less(i, j int) bool {
+	if l[i].Value == l[j].Value {
+		if l[i].SPDXExpression == l[j].SPDXExpression {
+			if l[i].Type == l[j].Type {
+				// While URL and location are not exclusive fields
+				// returning true here reduces the number of swaps
+				// while keeping a consistent sort order of
+				// the order that they appear in the list initially
+				// If users in the future have preference to sorting based
+				// on the slice representation of either field we can update this code
+				return true
+			}
+			return l[i].Type < l[j].Type
+		}
+		return l[i].SPDXExpression < l[j].SPDXExpression
+	}
+	return l[i].Value < l[j].Value
+}
+
+func (l Licenses) Swap(i, j int) {
+	l[i], l[j] = l[j], l[i]
+}
+
+func NewLicense(value string) License {
+	spdxExpression, err := license.ParseExpression(value)
+	if err != nil {
+		log.Trace("unable to parse license expression: %w", err)
+	}
+
+	// TODO: how do we express other places where a license is declared
+	// EX: we got this from the go module cache at path /x/y/z on disk
+	return License{
+		Value:          value,
+		SPDXExpression: spdxExpression,
+		Type:           license.Declared,
+		URL:            internal.NewStringSet(),
+		Location:       source.NewLocationSet(),
+	}
+}
+
+func NewLicenseFromType(value string, t license.Type) License {
+	spdxExpression, err := license.ParseExpression(value)
+	if err != nil {
+		log.Trace("unable to parse license expression: %w", err)
+	}
+
+	return License{
+		Value:          value,
+		SPDXExpression: spdxExpression,
+		Type:           t,
+		URL:            internal.NewStringSet(),
+		Location:       source.NewLocationSet(),
+	}
+}
+
+func NewLicensesFromValues(values ...string) (licenses []License) {
+	for _, v := range values {
+		// ignore common SPDX license expression connectors
+		// that could be included in input
+		if v == "" || v == "AND" {
```

Comment:
I'm surprised to see `AND` processing here, is this the right spot for this? And if so, shouldn't this also consider other keywords such as `WITH` and `OR`?